Update package article samples (#154)
hubverse-org/hubEnsembles: vignettes/articles/hubEnsembles.Rmd

# Creating ensembles with `linear_pool`

> The `linear_pool()` function implements the linear opinion pool (LOP, also known as a distributional mixture) method when ensembling predictions. This function can be used to combine predictions with output types mean, CDF, PMF, sample, and quantile. Unlike `simple_ensemble()`, this function handles its computation differently based on the output type. For the CDF, PMF, and mean output types, the linear pool method is equivalent to calling `simple_ensemble()` with a mean aggregation function, since `simple_ensemble()` produces a linear pool prediction (an average of individual model cumulative or bin probabilities).

Author: 1. From what I understand, LOP is well-known enough not to warrant a citation.
2. LOP averages probabilities at the same $x$ value to create the ensemble distribution. Since the median output type has a fixed probability value (0.5), it is incompatible with LOP. Conversely, the mean output type isn't tied to a fixed probability value, so it doesn't suffer the same type of incompatibility issue

> The `linear_pool()` function implements the linear opinion pool (LOP, also known as a distributional mixture) method when ensembling predictions. This function can be used to combine predictions with output types mean, CDF, PMF, sample, and quantile. Unlike `simple_ensemble()`, this function handles its computation differently based on the output type. For the CDF, PMF, and mean output types, the linear pool method is equivalent to calling `simple_ensemble()` with a mean aggregation function, since `simple_ensemble()` produces a linear pool prediction (an average of individual model cumulative or bin probabilities).

Reviewer: Thank you for answering!

> The `linear_pool()` function implements the linear opinion pool (LOP, also known as a distributional mixture) method when ensembling predictions. This function can be used to combine predictions with output types mean, CDF, PMF, sample, and quantile. Unlike `simple_ensemble()`, this function handles its computation differently based on the output type. For the CDF, PMF, and mean output types, the linear pool method is equivalent to calling `simple_ensemble()` with a mean aggregation function, since `simple_ensemble()` produces a linear pool prediction (an average of individual model cumulative or bin probabilities).

Reviewer 2: You could consider using the [@stone1961] and [@lichtendahl2013] citations from the hub ensembles manuscript here

> For the sample output type, the LOP method pools the input sample predictions into a combined ensemble distribution. By default, the `linear_pool()` function will simply collect and return all provided samples. However, the user may also specify a number of sample predictions for the ensemble to return using the `n_output_samples` argument, in which case a random subset of predictions from individual models will be selected to create the linear pool of samples so that all component models are represented equally. This random selection of samples is stratified by model so that approximately the same number of samples from each individual model is included in the ensemble.

Reviewer 3: I think the text answers this question already. E.g. the next sentence answers Zhian's specific question, albeit without the specifics of 5x100=500:
> By default, the `linear_pool()` function will simply collect and return all provided samples.
And then as to the "how does it choose which samples to include"
> ... a random subset of predictions from individual models will be selected to create the linear pool of samples so that all component models are represented equally. This random selection of samples is stratified by model so that approximately the same number of samples from each individual model is included in the ensemble.

> For the sample output type, the LOP method pools the input sample predictions into a combined ensemble distribution. By default, the `linear_pool()` function will simply collect and return all provided samples. However, the user may also specify a number of sample predictions for the ensemble to return using the `n_output_samples` argument, in which case a random subset of predictions from individual models will be selected to create the linear pool of samples so that all component models are represented equally. This random selection of samples is stratified by model so that approximately the same number of samples from each individual model is included in the ensemble.

Reviewer: I was just taken aback because the sample handling is different from the other output types since the size of the sample ensemble model output is the same size as the other models combined (as opposed to mean, which returns a single value).

> For the sample output type, the LOP method pools the input sample predictions into a combined ensemble distribution. By default, the `linear_pool()` function will simply collect and return all provided samples. However, the user may also specify a number of sample predictions for the ensemble to return using the `n_output_samples` argument, in which case a random subset of predictions from individual models will be selected to create the linear pool of samples so that all component models are represented equally. This random selection of samples is stratified by model so that approximately the same number of samples from each individual model is included in the ensemble.

Reviewer 2: Maybe we could clarify by stating this directly?

> "By default, the `linear_pool()` function will simply collect and return all provided samples, so that the number of samples for the ensemble is equal to the sum of the number of samples from all individual models."

The sample output type is unique in that it may represent predictions from either joint or marginal predictive distributions, unlike the other output types which only ever summarize the latter. Hence, sample representation may capture dependence across combinations of multiple modeled task ID variables when recording samples from a joint predictive distribution. In this case, samples with the same index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple levels of the task id variables.

When requesting a linear pool made up of a subset of the input sample predictions, the user must identify the dependence structure using the `compound_taskid_set` parameter to ensure the resulting ensemble is valid. The compound task ID set consists of task ID variables that do not display dependence amongst their values, and thus each unique combination of their values can be seen as representations of a single multivariate outcome of interest (or compound modeling task). If no task ID variables display dependents, then the `compound_taskid_set` consists of all task ID variables.

> The example data for the sample output type has task ID variables "reference_date", "location", "horizon", "target", and "target_end_date"; of these five, only "horizon" displays dependence while "target_end_date" is derived from a combination of "reference_date" and "horizon". The other three task ID variables thus make up the compound task ID set, which is specified in the call to `linear_pool()`.

Reviewer 2: ```suggestion
The example data for the sample output type has task ID variables `"reference_date"`, `"location"`, `"horizon"`, `"target"`, and `"target_end_date"`. In this example, the samples capture dependence across different forecast `"horizon"`s. However, the samples do not capture dependence across different `"reference_date"`s, `"location"`s, or `"target"`s. Thus, those three task id variables make up the compound task ID set that is specified in the call to `linear_pool()`. The `"target_end_date"` for a given forecast is derived from the combination of `"reference_date"` and `"horizon"`, and so it is specified as the argument for `derived_tasks`.
```

```{r}
hubExamples::forecast_outputs |>
  dplyr::filter(output_type == "sample") |>
  dplyr::mutate(output_type_id = as.numeric(output_type_id)) |> # make indices numeric for readability
  hubEnsembles::linear_pool(
    weights = NULL,
    model_id = "linear-pool-joint",
    task_id_cols = c("reference_date", "location", "horizon", "target", "target_end_date"),
    compound_taskid_set = c("reference_date", "location", "target"),
    derived_tasks = "target_end_date",
    n_output_samples = 100
  )
```

> Derived task IDs generally are not part of the compound task ID set **unless** all the task ID variables their values depend upon are already part of the compound task ID set. The `derived_tasks` argument is used to validate the provided compound task ID set is compatible with the input sample predictions. If there are no derived tasks, then the argument may be set to `NULL`.

Reviewer 2: ```suggestion
Derived task IDs generally are not part of the compound task ID set **unless** all the task ID variables their values depend upon are already part of the compound task ID set. The `derived_tasks` argument is used to validate that the provided compound task ID set is compatible with the input sample predictions. If there are no derived tasks, then the argument may be set to `NULL`.
```

head(weighted_linear_pool_norm, 10)
```

## Changing the parametric family used for extrapolation into distribution tails

We can also change the distribution that `distfromq` uses to approximate the tails of component models' predictive distributions to either log normal or Cauchy using the `tail_dist` argument. This choice usually does not have a large impact on the resulting ensemble distribution, though, and can only be seen in its outer edges. (For more details and other function options, see the documentation in the [`distfromq` package](https://reichlab.io/distfromq/).)

```{r}
linear_pool_lnorm <- hubExamples::forecast_outputs |>
  dplyr::filter(output_type %in% "quantile") |>
  hubEnsembles::linear_pool(
    model_id = "linear-pool-lognormal",
    tail_dist = "lnorm"
  )
linear_pool_cauchy <- hubExamples::forecast_outputs |>
  dplyr::filter(output_type %in% "quantile") |>
  hubEnsembles::linear_pool(
    model_id = "linear-pool-cauchy",
    tail_dist = "cauchy"
  )
```

## Ensembling samples from joint distributions

> The sample output type is unique in that it may represent predictions from either joint or marginal predictive distributions, unlike the other output types which only ever summarize the latter. Hence, the sample representation may capture dependence across combinations of multiple modeled task ID variables when recording samples from a joint predictive distribution. In this case, samples with the same index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple levels of the task id variables.

Reviewer 2: I'm definitely open to continuing iteration on this!  That said, I find "all the variables in each model task" to be unclear.  I'd understand "all the variables" to be "all the columns", i.e. location, reference date, horizon, target date, and target in this example.  But the "jointness" is not across those columns, it's cross different settings of those variables represented across multiple rows.

> The sample output type is unique in that it may represent predictions from either joint or marginal predictive distributions, unlike the other output types which only ever summarize the latter. Hence, the sample representation may capture dependence across combinations of multiple modeled task ID variables when recording samples from a joint predictive distribution. In this case, samples with the same index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple levels of the task id variables.

Reviewer: It's a tough cookie because we are asking people to consider three layers of relationships and that's like describing the path of a rumor, which feels awkward ("My father's boss's kid's friend said that Marmite is better than Vegemite").

Part of me was going off of the [definition of a model task](https://hubverse.io/en/latest/user-guide/tasks.html) as being the unique combination of values across task ID variables, so once you have that definition, then there can only be one combination of values per model task. Although, the documentation we have right now conflates "model task" and "target" 🙁:

> The [task_ids object](https://hubverse.io/en/latest/user-guide/tasks.html#task-id-vars) defines both labels for columns in submission files and the set of valid values for each column. Any unique value combination defines a single modeling task or target.

What about "unique value combinations among the modeling task ID variables"

> The sample output type is unique in that it may represent predictions from either joint or marginal predictive distributions, unlike the other output types which only ever summarize the latter. Hence, the sample representation may capture dependence across combinations of multiple modeled task ID variables when recording samples from a joint predictive distribution. In this case, samples with the same index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple levels of the task id variables.

Reviewer 2: I definitely agree with you that this is hard to say clearly.

1. In this case, samples with the same index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple levels of the task id variables.
2. In this case, samples with the same index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple unique value combinations among the modeling task ID variables.

Not sure that what I wrote under point 2 is what you were suggesting.  What I wrote still doesn't feel clearer to me though.

What about:

3. In this case, samples with the same index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple combinations of values for the task ID variables.

When requesting a linear pool made up of a subset of the input sample predictions, the user must identify the dependence structure using the `compound_taskid_set` parameter to ensure the resulting ensemble is valid. The compound task ID set consists of task ID variables that, together, identify a "compound modeling task" corresponding to a single modeled unit with a multivariate outcome of interest. Samples are generally not expected to capture dependence across different compound tasks, but are expected to capture dependence across different outcomes pertaining to the same compound task. If samples are not expected to capture dependence across the levels of any task ID variables, the `compound_taskid_set` consists of all task ID variables. Please see the [hubverse documentation on samples](https://hubverse.io/en/latest/user-guide/sample-output-type.html) for more information about compound tasks.

> The example data for the sample output type has task ID variables `"reference_date"`, `"location"`, `"horizon"`, `"target"`, and `"target_end_date"`. In this example, the samples capture dependence across different forecast `"horizon"`s. However, the samples do not capture dependence across different `"reference_date"`s, `"location"`s, or `"target"`s. Thus, those three task id variables make up the compound task ID set that is specified in the call to `linear_pool()`. The `"target_end_date"` for a given forecast is derived from the combination of `"reference_date"` and `"horizon"`, and so it is specified as the argument for `derived_tasks`.

Reviewer: I was referring to the entire section, because you are introducing two concepts here: the compound task ID set and the derived task IDs.

Below is my suggestion (emphasis are my additions).

> All output types will summarize predictions from marginal distributions. The sample output type is unique in that **it can additionally represent predictions from joint predictive distributions**. Hence, the sample representation may capture dependence across combinations of multiple modeled task ID variables when recording samples from a joint predictive distribution. In this case, samples with the same index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple combinations of values for the task ID variables.
> 
> The example data for the sample output type has task ID variables `"reference_date"`, `"location"`, `"horizon"`, `"target"`, and `"target_end_date"`. In this example, the samples capture dependence across different forecast `"horizon"`s. However, the samples do not capture dependence across different `"reference_date"`s, `"location"`s, or `"target"`s.
> 
> When requesting a linear pool made up of a subset of the input sample predictions, the user must identify the dependence structure using the `compound_taskid_set` parameter to ensure the resulting ensemble is valid. The compound task ID set consists of task ID variables that, together, identify a "compound modeling task" corresponding to a single modeled unit with a multivariate outcome of interest. Samples are generally not expected to capture dependence across different compound tasks, but are expected to capture dependence across different outcomes pertaining to the same compound task. If samples are not expected to capture dependence across the levels of any task ID variables, the `compound_taskid_set` consists of all task ID variables. Please see the [hubverse documentation on samples](https://hubverse.io/en/latest/user-guide/sample-output-type.html) for more information about compound tasks.
> 
> Thus, *in this example*, those three task id variables *(`"reference_date"`, `"location"`, and `"target"`)* make up the compound task ID set that is specified in the call to `linear_pool()`. The `"target_end_date"` for a given forecast is derived from the combination of `"reference_date"` and `"horizon"`, and so it is specified as the argument for `derived_tasks`.

This way you have a nice pairing of explanation, example, explanation, example

> The example data for the sample output type has task ID variables `"reference_date"`, `"location"`, `"horizon"`, `"target"`, and `"target_end_date"`. In this example, the samples capture dependence across different forecast `"horizon"`s. However, the samples do not capture dependence across different `"reference_date"`s, `"location"`s, or `"target"`s. Thus, those three task id variables make up the compound task ID set that is specified in the call to `linear_pool()`. The `"target_end_date"` for a given forecast is derived from the combination of `"reference_date"` and `"horizon"`, and so it is specified as the argument for `derived_tasks`.

Author: That makes sense, thanks for the clarification! Also, does this re-phrasing take care of your request for a minimal example in the previous comment chain?

> The example data for the sample output type has task ID variables `"reference_date"`, `"location"`, `"horizon"`, `"target"`, and `"target_end_date"`. In this example, the samples capture dependence across different forecast `"horizon"`s. However, the samples do not capture dependence across different `"reference_date"`s, `"location"`s, or `"target"`s. Thus, those three task id variables make up the compound task ID set that is specified in the call to `linear_pool()`. The `"target_end_date"` for a given forecast is derived from the combination of `"reference_date"` and `"horizon"`, and so it is specified as the argument for `derived_tasks`.

Reviewer: > Also, does this re-phrasing take care of your request for a minimal example in the previous comment chain?

I am still confused by this particular sentence.

> Samples are generally not expected to capture dependence across different compound tasks, but are expected to capture dependence across different outcomes pertaining to the same compound task. 

Mainly because you have the fragment "expected to capture dependence across different" repeated twice and it requires a bit of effort to parse. A simplification might be:

> Samples are expected to capture dependence across different outcomes within the same compound task (not across compound tasks). 

That being said, I still find the concept hard to grok, but that might be because IANAS (I am not a statistician). 

An example that could show two compound tasks and how the samples would capture dependence across horizons would really help.

```{r}
hubExamples::forecast_outputs |>
  dplyr::filter(output_type == "sample") |>
  dplyr::mutate(output_type_id = as.numeric(output_type_id)) |> # make indices numeric for readability
  hubEnsembles::linear_pool(
    weights = NULL,
    model_id = "linear-pool-joint",
    task_id_cols = c("reference_date", "location", "horizon", "target", "target_end_date"),
    compound_taskid_set = c("reference_date", "location", "target"),

>     derived_tasks = "target_end_date",

Reviewer: Done: https://github.com/hubverse-org/hubEnsembles/issues/156

    n_output_samples = 100
  )
```

> Derived task IDs generally are not part of the compound task ID set **unless** all the task ID variables their values depend upon are already part of the compound task ID set. The `derived_tasks` argument is used to validate that the provided compound task ID set is compatible with the input sample predictions. If there are no derived tasks, then the argument may be set to `NULL`.

Author: I expanded the ideas in this paragraph and moved it above the example code to better integrate with the previous paragraphs; let me know if it's clearer now.

head(weighted_linear_pool_norm, 10)
```

## Changing the parametric family used for extrapolation into distribution tails

We can also change the distribution that `distfromq` uses to approximate the tails of component models' predictive distributions to either log normal or Cauchy using the `tail_dist` argument. This choice usually does not have a large impact on the resulting ensemble distribution, though, and can only be seen in its outer edges. (For more details and other function options, see the documentation in the [`distfromq` package](https://reichlab.io/distfromq/).)

```{r}
linear_pool_lnorm <- hubExamples::forecast_outputs |>
  dplyr::filter(output_type == "quantile") |>
  hubEnsembles::linear_pool(
    model_id = "linear-pool-lognormal",
    tail_dist = "lnorm"
  )
linear_pool_cauchy <- hubExamples::forecast_outputs |>
  dplyr::filter(output_type == "quantile") |>
  hubEnsembles::linear_pool(
    model_id = "linear-pool-cauchy",
    tail_dist = "cauchy"
  )
```

## Ensembling samples from joint distributions

> All output types will summarize predictions from marginal distributions. The sample output type is unique in that **it can additionally represent predictions from joint predictive distributions**. Hence, the sample representation may capture dependence across combinations of multiple modeled task ID variables when recording samples from a joint predictive distribution. In this case, samples with the same index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple combinations of values for the task ID variables.

Reviewer 2: ```suggestion
All output types will summarize predictions from marginal distributions. The sample output type is unique in that **it can additionally represent predictions from joint predictive distributions**. Hence, the sample representation may capture dependence across multiple combinations of values for modeled task ID variables when recording samples from a joint predictive distribution. In this case, samples with the same index (specified by the `output_type_id`) may be assumed to correspond to a single sample from a joint distribution across multiple combinations of values for the task ID variables.
```

The example data for the sample output type has task ID variables `"reference_date"`, `"location"`, `"horizon"`, `"target"`, and `"target_end_date"`. In this example, the samples capture dependence across different forecast `"horizon"`s. However, the samples do not capture dependence across different `"reference_date"`s, `"location"`s, or `"target"`s.

> When specifically requesting a linear pool made up of a **subset of the input sample predictions**, the user must identify the dependence structure using the `compound_taskid_set` parameter to ensure the resulting ensemble is valid. The compound task ID set consists of task ID variables that, together, identify a "compound modeling task" corresponding to a single modeled unit with a multivariate outcome of interest. Samples are expected to capture dependence across different outcomes within the same compound task (not across compound tasks), as each compound task is assumed to be independent of all other compound tasks. 

Author: I think one unique consideration for hubEnsembles is that we're combining predictions from multiple models. So there's some sense in which we shouldn't be allowing users to specify more dependence to capture then the minimum level specified by the hub (because I'm not sure the resulting ensemble would be valid if component models capture different amounts of dependence)

> When specifically requesting a linear pool made up of a **subset of the input sample predictions**, the user must identify the dependence structure using the `compound_taskid_set` parameter to ensure the resulting ensemble is valid. The compound task ID set consists of task ID variables that, together, identify a "compound modeling task" corresponding to a single modeled unit with a multivariate outcome of interest. Samples are expected to capture dependence across different outcomes within the same compound task (not across compound tasks), as each compound task is assumed to be independent of all other compound tasks. 

Reviewer 2: > I'm not sure the resulting ensemble would be valid if component models capture different amounts of dependence

I think the answer is that it is a little odd, but it is valid as far as capturing dependence at the minimal level. Take our example with task ids location and horizon, where location is in the compound task id set but horizon is not.  The hub is saying, "at minimum, you should model dependence across forecast horizons."  Now suppose model A only models temporal dependence, while model B models dependence over both time and spatial units.  Suppose there are 4 horizons and 3 states predicted, and the hub asks for 10 samples.  We will have validated that both models submitted predictions for all 4 horizons.

For the ensemble, we will draw 5 samples for each model at random within each location (i.e., within the compound units specified by the compound task id set).  For model A, this is of course fine: we will select 5 of their submitted trajectories for each location.  For model B, we may do something like select samples 1 through 5 for the first location, samples 3 through 7 for the second location, and samples 6 through 10 for the third location.  This would be a little funny, because for that model there are some overlapping/shared samples for locations 1 and 2, and some shared samples for locations 2 and 3, but no shared samples for that model between locations 1 and 3 or for the other model between any pairs of locations.

However, if you just consider the sample trajectories within each location (as specified by the compound task id set the analyst provided), these are valid samples from the "marginal joint" distributions across horizons separately within locations.

> When specifically requesting a linear pool made up of a **subset of the input sample predictions**, the user must identify the dependence structure using the `compound_taskid_set` parameter to ensure the resulting ensemble is valid. The compound task ID set consists of task ID variables that, together, identify a "compound modeling task" corresponding to a single modeled unit with a multivariate outcome of interest. Samples are expected to capture dependence across different outcomes within the same compound task (not across compound tasks), as each compound task is assumed to be independent of all other compound tasks. 

Author: Okay, that makes sense. So I should essentially just go with implementing the change at the top of this thread?

> When specifically requesting a linear pool made up of a **subset of the input sample predictions**, the user must identify the dependence structure using the `compound_taskid_set` parameter to ensure the resulting ensemble is valid. The compound task ID set consists of task ID variables that, together, identify a "compound modeling task" corresponding to a single modeled unit with a multivariate outcome of interest. Samples are expected to capture dependence across different outcomes within the same compound task (not across compound tasks), as each compound task is assumed to be independent of all other compound tasks. 

Reviewer 2: yeah, i think so

> For example, a compound task could be predicting the number of weekly incident influenza hospitalizations (`"target"`) in Massachusetts (`"location"`) starting on November 19, 2022 (`"reference_date"`). Here, `"horizon"` is not part of the compound task ID set, indicating that sample predictions made at each horizon depend on those for the previous horizon (i.e., 2 week ahead forecasts depend on 1 week ahead forecasts) within every compound task for the sample output type. Conversely, dependence across any compound tasks (such as a relationship between targets like influenza hospitalizations and deaths) would not be captured based on the current specification of the compound task ID set.

Reviewer 2: see the suggestion/question immediately preceding this one.
```suggestion
For example, a compound task could be predicting the number of weekly incident influenza hospitalizations (`"target"`) in Massachusetts (`"location"`) starting on November 19, 2022 (`"reference_date"`). Here, `"horizon"` is not part of the compound task ID set, indicating that sample predictions made at each horizon depend on those for the previous horizon (i.e., 2 week ahead forecasts depend on 1 week ahead forecasts) within every compound task for the sample output type.
```

> For example, a compound task could be predicting the number of weekly incident influenza hospitalizations (`"target"`) in Massachusetts (`"location"`) starting on November 19, 2022 (`"reference_date"`). Here, `"horizon"` is not part of the compound task ID set, indicating that sample predictions made at each horizon depend on those for the previous horizon (i.e., 2 week ahead forecasts depend on 1 week ahead forecasts) within every compound task for the sample output type. Conversely, dependence across any compound tasks (such as a relationship between targets like influenza hospitalizations and deaths) would not be captured based on the current specification of the compound task ID set.

Reviewer 2: This is a minor suggestion.  The original phrasing implies a specific dependence structure, basically an AR(1) process where the distribution at time t depends on the past only through the data at time t-1.  But dependence across time points could be more complex.
```suggestion
For example, a compound task could be predicting the number of weekly incident influenza hospitalizations (`"target"`) in Massachusetts (`"location"`) starting on November 19, 2022 (`"reference_date"`). Here, `"horizon"` is not part of the compound task ID set, indicating that sample predictions made at each horizon depend on those for the other horizons within every compound task for the sample output type. Each sample can therefore be interpreted as a trajectory giving a possible path of hospitalizations over time.
```

Thus, in this example, those three task id variables (`"reference_date"`, `"location"`, and `"target"`) make up the compound task ID set that is specified in the call to `linear_pool()`. The `"target_end_date"` for a given forecast is derived from the combination of `"reference_date"` and `"horizon"`, and so it is specified as the argument for `derived_tasks`.

```{r}
hubExamples::forecast_outputs |>
  dplyr::filter(output_type == "sample") |>
  dplyr::mutate(output_type_id = as.numeric(output_type_id)) |> # make indices numeric for readability
  hubEnsembles::linear_pool(
    weights = NULL,
    model_id = "linear-pool-joint",
    task_id_cols = c("reference_date", "location", "horizon", "target", "target_end_date"),
    compound_taskid_set = c("reference_date", "location", "target"),
    derived_tasks = "target_end_date",
    n_output_samples = 100
  )
```

> Derived task IDs are another subset of task ID variables whose values are "derived" from a combination of the values from other task ID variables, which may or may not be part of the compound task ID set. In the example, the derived task ID `"target_end_date"` is not part of the compound task ID set because `"reference_date"` and `"horizon"` are not both part of the compound task ID set.  

Reviewer 2: These sentences are slightly redundant with text in the last sentence just before the code block, and expands on ideas that you started to introduce there.  How about removing it from here and combining with the previous paragraph?

Reviewer 2: I take this suggestion back.
